# Solução de sistemas lineares através da eliminação de Gauss

## License

All content can be freely used and adapted under the terms of the 
[Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).

![Creative Commons License](https://i.creativecommons.org/l/by/4.0/88x31.png)

## Imports

Coloque **todos** os `import` na célula abaixo. Não se esqueça do `%matplotlib inline` para que os gráficos apareçam no notebook.

In [34]:
import numpy as np

## Tarefa

Reduza o sistema linear $\bar{\bar{A}}\bar{x} = \bar{b}$ para sua forma escalonada (row echelon form).

In [35]:
N = 11  # Número de linhas
M = 11  # Número de colunas
sistema = [[25, 108, 58, 102, 61, 24, 39, 40, 67, 11, 5],
           [51, 28, 17, 98, 105, 43, 86, 2, 71, 20, 95],
           [56, 70, 109, 45, 117, 52, 97, 92, 1, 42, 78],
           [59, 6, 50, 37, 120, 118, 104, 64, 53, 100, 66],
           [82, 74, 68, 80, 16, 115, 63, 103, 60, 30, 93],
           [79, 113, 65, 72, 22, 35, 101, 10, 83, 89, 85],
           [34, 57, 90, 9, 14, 47, 76, 44, 26, 48, 114],
           [41, 7, 21, 75, 33, 49, 32, 119, 46, 0, 87],
           [91, 77, 55, 18, 106, 54, 116, 12, 13, 31, 38],
           [27, 81, 29, 107, 84, 15, 3, 112, 88, 69, 19],
           [110, 96, 23, 73, 36, 8, 99, 4, 94, 111, 62]]
lado_direito = [3323, 2869, 4085, 3211, 4181, 3831, 2699, 2891, 3188, 3586, 3598]

In [36]:
for k in range(N - 1): # ler todas as linhas da matriz
    for i in range(k + 1, N): # Para cada linha da matriz, fazer:
        valor = -sistema[i][k]/sistema[k][k] # valor pelo qual a linha é multiplicada
        lado_direito[i] = lado_direito[i] + valor*lado_direito[k]# como é um sistema os "resultados" também são multiplicados
        for j in range(k, M):# movimenta a soma entre as colunas das linahs que estão sendo somadas
            sistema[i][j] = sistema[i][j] + valor*sistema[k][j] #soma entre as linhas k e i, e o valor obtido sendo substituido na posição da linha i
print(sistema)#imprime a matriz

[[25, 108, 58, 102, 61, 24, 39, 40, 67, 11, 5], [0.0, -192.32, -101.32000000000001, -110.08000000000001, -19.439999999999998, -5.960000000000001, 6.439999999999998, -79.6, -65.68, -2.4400000000000013, 84.8], [-7.105427357601002e-15, 0.0, 69.6526622296173, -85.0765391014975, -2.2620632279534263, 3.56780366056572, 3.883111480865212, 73.55657237936771, -90.36688851913478, 19.541181364392678, -9.004991680532456], [0.0, 0.0, 0.0, -7.232751052584419, 2.6338439487593206, 66.80682612201022, 1.1598136701603692, 25.89286034220192, 37.269497446922884, 64.78667024993281, -49.819887127116345], [0.0, 0.0, 0.0, 0.0, -177.9183147715115, -539.4790163885706, -85.8438255519006, -165.04626748840752, -356.4192535967665, -575.1593971701487, 391.184215350165], [0.0, 0.0, 0.0, 0.0, -2.842170943040401e-14, -539.655975046652, 43.02969123025166, -267.1570084206994, -271.02468416529314, -378.2996057969407, 357.87920483633803], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 55.46848311433081, -69.824450898192, 59.86020982430274, 

### Resultado esperado

Quando executada, a célula abaixo deverá imprimir exatamente:

                                                   A                                                   |    b 
      25.000  108.000   58.000  102.000   61.000   24.000   39.000   40.000   67.000   11.000    5.000 | 3323.000
       0.000 -192.320 -101.320 -110.080  -19.440   -5.960    6.440  -79.600  -65.680   -2.440   84.800 | -3909.920
      -0.000    0.000   69.653  -85.077   -2.262    3.568    3.883   73.557  -90.367   19.541   -9.005 |  136.662
       0.000    0.000    0.000   -7.233    2.634   66.807    1.160   25.893   37.269   64.787  -49.820 |  341.724
       0.000    0.000    0.000    0.000 -177.918 -539.479  -85.844 -165.046 -356.419 -575.159  391.184 | -4053.756
       0.000    0.000    0.000    0.000   -0.000 -539.656   43.030 -267.157 -271.025 -378.300  357.879 | -3235.329
       0.000    0.000    0.000    0.000    0.000    0.000   55.468  -69.824   59.860   57.791   78.381 |  141.706
       0.000    0.000    0.000    0.000    0.000    0.000    0.000   70.381   24.471   60.345  -29.950 |  741.794
       0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000  -16.843 -126.224   62.584 |   57.797
       0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000 -291.983  268.919 |  537.838
      -0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000  -71.547 | -143.094


In [37]:
print('                                               A                                                   |    b ')
for i in range(N):
    for j in range(M):
        print('{:8.3f} '.format(sistema[i][j]), end='')
    print('| {:8.3f}'.format(lado_direito[i]))

                                               A                                                   |    b 
  25.000  108.000   58.000  102.000   61.000   24.000   39.000   40.000   67.000   11.000    5.000 | 3323.000
   0.000 -192.320 -101.320 -110.080  -19.440   -5.960    6.440  -79.600  -65.680   -2.440   84.800 | -3909.920
  -0.000    0.000   69.653  -85.077   -2.262    3.568    3.883   73.557  -90.367   19.541   -9.005 |  136.662
   0.000    0.000    0.000   -7.233    2.634   66.807    1.160   25.893   37.269   64.787  -49.820 |  341.724
   0.000    0.000    0.000    0.000 -177.918 -539.479  -85.844 -165.046 -356.419 -575.159  391.184 | -4053.756
   0.000    0.000    0.000    0.000   -0.000 -539.656   43.030 -267.157 -271.025 -378.300  357.879 | -3235.329
   0.000    0.000    0.000    0.000    0.000    0.000   55.468  -69.824   59.860   57.791   78.381 |  141.706
   0.000    0.000    0.000    0.000    0.000    0.000    0.000   70.381   24.471   60.345  -29.950 |  741.794
   0.000  

## Tarefa

Utilize a matriz escalonada acima para encontrar a solução do sistema através da substituição retrocedida (back substitution). Coloque a solução em uma lista chamada `solucao`.

**Dica**: 

* Para criar uma lista com um mesmo valor repetido N vezes: `[numero]*N`. Por exemplo, `[0]*4` resulta em `[0, 0, 0, 0]`.

In [38]:
#Pensar de trás pra frente
solucao = [0]*M #fazer lista de zeros para depois ir substituindo os valores, mais fácil que usar o append
soma = 0
print(lado_direito)
for k in range(N-1 , -1, -1): # já que a última posição no caso é zero, devemos ir até a posição -1, porque deve contar a posição zero também # se é de trás pra frente, então deve-se ir de -1 em -1, porque está retrocedendo uma posição
    soma = 0
    for i in range(k + 1, N):
        soma = soma + (sistema[k][i]*solucao[i])
    resultado = (lado_direito[k] - soma)/sistema[k][k]
    solucao[k] = resultado
print(solucao)           

[3323, -3909.92, 136.66222961730455, 341.7244289169582, -4053.75582368912, -3235.3288564001277, 141.70569019401557, 741.7935097847427, 57.79745228918122, 537.8381271958178, -143.09370350431925]
[5.999999999999891, 9.000000000000153, 4.99999999999979, 7.000000000000029, 2.9999999999999796, 1.0000000000000544, 8.000000000000094, 10.000000000000036, 3.999999999999848, 5.0227639466607155e-14, 2.0000000000000404]


### Resultado esperado

Quando executada, a célula abaixo deverá imprimir exatamente:

    6.000000 9.000000 5.000000 7.000000 3.000000 1.000000 8.000000 10.000000 4.000000 0.000000 2.000000 

In [39]:
for i in range(N):
    print('{:8.6f} '.format(solucao[i]), end='')

6.000000 9.000000 5.000000 7.000000 3.000000 1.000000 8.000000 10.000000 4.000000 0.000000 2.000000 

## Tarefa Bônus

Sua tarefa será aplicar a eliminação de Gauss para resolver o sistema linear gerado abaixo.

In [29]:
# Para tornar a vida mais interessante, vamos gerar um sistema linear aleatório
N, M = 13, 13
A = np.random.random_integers(10, 50, size=(N, M)).tolist()
for i in range(N):
    A[i][i] = 0
b = np.dot(A, np.random.random_integers(1, 20, size=N)).tolist()

In [30]:
print('                                                   A                                                    |    b ')
for i in range(N):
    for j in range(M):
        print('{:7.1f} '.format(A[i][j]), end='')
    print('| {:7.1f}'.format(b[i]))

                                                   A                                                    |    b 
    0.0    49.0    24.0    13.0    13.0    26.0    14.0    27.0    30.0    22.0    27.0    37.0    24.0 |  3502.0
   14.0     0.0    35.0    36.0    36.0    10.0    42.0    20.0    15.0    47.0    36.0    24.0    33.0 |  3433.0
   33.0    37.0     0.0    26.0    44.0    34.0    12.0    25.0    15.0    42.0    31.0    30.0    28.0 |  3473.0
   11.0    18.0    14.0     0.0    11.0    37.0    36.0    27.0    26.0    43.0    37.0    26.0    48.0 |  3010.0
   17.0    31.0    25.0    45.0     0.0    30.0    17.0    12.0    27.0    45.0    24.0    28.0    38.0 |  3766.0
   15.0    18.0    14.0    32.0    10.0     0.0    25.0    45.0    37.0    47.0    33.0    47.0    46.0 |  3931.0
   23.0    12.0    28.0    47.0    17.0    30.0     0.0    30.0    16.0    17.0    14.0    18.0    28.0 |  3294.0
   26.0    28.0    43.0    28.0    26.0    33.0    33.0     0.0    13.0    24.0    41.0   

Porém, há um problema. Você não poderá fazer isso diretamente pois $A_{ii} == 0$ (a diagonal da matriz é zero). Logo, não há como dividir pelo elemento pivo e não há como aplicar a eliminação de Gauss nesse sistema. **Experimente aplicar a eliminação convencional que você implementou acima nesse sistema e veja o que acontece.**

In [31]:
#Pensar de trás pra frente
solucao = [0]*M #fazer lista de zeros para depois ir substituindo os valores, mais fácil que usar o append
soma = 0
print(lado_direito)
for k in range(N-1 , -1, -1): # já que a última posição no caso é zero, devemos ir até a posição -1, porque deve contar a posição zero também # se é de trás pra frente, então deve-se ir de -1 em -1, porque está retrocedendo uma posição
    soma = 0
    for i in range(k + 1, N):
        soma = soma + (sistema[k][i]*solucao[i])
    resultado = (lado_direito[k] - soma)/sistema[k][k]
    solucao[k] = resultado
print(solucao)        

[3323, -3909.92, 136.66222961730455, 341.7244289169582, -4053.75582368912, -3235.3288564001277, 141.70569019401557, 741.7935097847427, 57.79745228918122, 537.8381271958178, -143.09370350431925]


IndexError: list index out of range



### Condensação pivotal (partial pivoting)

Existe um método para contornar esse problema chamado **condensação pivotal** (partial pivoting). A ideia é que podemos trocar duas linhas do sistema de lugar sem alterar a solução. Logo, podemos tirar os elementos 0 da diagonal da matriz trocando por outra linha sem o 0 na posição do pivo. Um jeito melhor ainda de fazer isso é colocando a linha com o **maior elemento** possível na posição do pivo (veja [aqui uma ótima explicação disso](http://www.math.iitb.ac.in/~neela/partialpivot.pdf)). 

O método da eliminação de Gauss com condensação pivotal é:

    Para cada linha k do sistema:
        Ache a linha de A com o maior elemento na coluna k (só precisa considerar de k para baixo)
        Troque a linha k com a linha determinada acima (não esqueça de trocar o lado direito também)
        pivo = A[k][k]
        Escalonamento para zerar a coluna k (igual ao feita na eliminação convencional)
        
        
Resolva o sistema linear definido acima utilizando eliminação de Gauss com condensação pivotal. Coloque a solução em uma variável chamada `solucao`.

### Resultado esperado

Quando executada, a célula abaixo deverá imprimir `True`.

In [32]:
print(np.allclose(np.linalg.solve(A, b), solucao))

False
